chain/rust/src/assets.rs

```diff
@@ -0,0 +1,595 @@
+where T: num::CheckedAdd + num::CheckedSub + num::Zero + num::Bounded + Copy + Clone + ClampedSub {
+    fn clamped_sub(&self, rhs: &Self) -> Self {
+        let mut bundle = self.clone();
+        for (policy, rhs_assets) in rhs.iter() {
+            for (asset_name, rhs_amount) in rhs_assets.iter() {
+                match bundle.get_mut(policy) {
+                    Some(lhs_assets) => match lhs_assets.get_mut(asset_name) {
+                        Some(lhs_amount) => match lhs_amount.checked_sub(rhs_amount) {
+                            Some(new_lhs_amount) => {
+                                if new_lhs_amount.is_zero() {
+                                    lhs_assets.remove(asset_name);
+                                    if lhs_assets.is_empty() {
+                                        bundle.remove(policy);
+                                    }
+                                } else {
+                                    *lhs_amount = new_lhs_amount;
+                                }
+                            },
+                            None => {
+                                if T::min_value().is_zero() {
+                                    // if underflow and unsigned, remove
+                                    lhs_assets.remove(asset_name);
+                                    if lhs_assets.is_empty() {
+                                        bundle.remove(policy);
+                                    }
+                                } else {
+                                    // otherwise keep it as minimum
+                                    *lhs_amount = T::min_value();
+                                }
+                            }
+                        },
+                        None => {
+                            // asset name is missing from left hand side
+                            if !T::min_value().is_zero() {
+                                bundle.set(policy.clone(), asset_name.clone(), T::zero().clamped_sub(rhs_amount));
+                            }
+                        }
+                    },
+                    None => {
+                        // policy id missing from left hand side
+                        if !T::min_value().is_zero() {
+                            bundle.set(policy.clone(), asset_name.clone(), T::zero().clamped_sub(rhs_amount));
+                        }
+                    }
+                }
+            }
+        }
+        bundle
+    }
+}
+
+impl<T> AssetBundle<T>
+    where T: num::CheckedAdd + num::CheckedSub + num::Zero + num::Bounded + Copy + Clone,
+{
+    /// Set the value of policy_id:asset_name to value.
+    /// Returns the previous value, or None if it didn't exist
+    pub fn set(&mut self, policy_id: PolicyId, asset_name: AssetName, value: T) -> Option<T> {
+        self.0.entry(policy_id).or_default().insert(asset_name, value)
+    }
+
+    /// Get the value of policy_id:asset_name. If it doesn't exist, 0 is returned.
```

Comment:
Is it more useful for users to have this turn into 0 if it's not present, or have an Option<T> returned?

Reply:
I prefer Option. I feel like returning zero will cause at least one person to get it wrong

Reply:
What if we provide both and have the zero one like `get_or_zero()`? (or whatever name would be better)

Reply:
or `count()`, `asset_count()`, etc.

Reply:
What would the benefit of `get_or_zero` be over just using an `unwrap_or` ?

counts might be useful to calculate the min utxo related to this bundle, but I'm not sure how many utility functions we want to expose given you can also just deref the type to get the underlying map and call utility functions on that, right?

Reply:
Yeah it would be more for convenience + there would be no `unwrap_or` in WASM and having it only in the WASM wrapper might be weird. We can leave it as just `get() -> Option<T>` if we want to limit the utility functionality though. For Cardano semantics is there a reason you'd ever have an explicit 0 instead of not having it in a multiasset/mint, and if you did would it matter (besides CBOR serialization details)? I guess for wallets/explorers/etc you might want to show it even if it's 0 maybe for the tx details, although you'd probably be iterating the object there instead of checking a specific policy id / asset name combo.

Reply:
I believe the only time you're ever allowed a 0 value of anything is when you're withdrawing 0 ADA from a reward account

Reply:
@SebastienGllmt Is there anything else here that needs changing before we can merge this? Or does it need more reviewing?

So should we just switch it to returning an `Option<T>`? Or if there cannot be a 0 value for any of these (since these are assets not ADA) keep it how it is or what? I don't have a strong preference. I guess what I said earlier about WASM not having `unwrap_or` isn't true since you can use null-coalescing in JS like `foo.get(id, name) ?? 0` if they need that so I guess the `Option<T>` is fine. The only possible issue is from what I can tell `None` is `undefined`, not `null` so if anyone directly does arithmetic on it you'd get NaN instead of 0 though with `null`, but only when the asset is absent. I guess lots of static analysis/type checking things would find that kind of issue though.

Reply:
yeah, returning either `undefined` or `null` in JS is totally fine. Almost everybody now-a-days uses Typescript or eslint which would warn about invalid arithmetic on possible undefined values